File: EaiGateway/src/main/java/com/yung/cho/eaigateway/config/RoutingConfig.java
package com.yung.cho.eaigateway.config;

import io.github.resilience4j.bulkhead.BulkheadConfig;
import io.github.resilience4j.bulkhead.BulkheadRegistry;
import io.github.resilience4j.circuitbreaker.CircuitBreakerConfig;
import io.github.resilience4j.circuitbreaker.CircuitBreakerRegistry;
import io.github.resilience4j.timelimiter.TimeLimiterConfig;
import io.github.resilience4j.timelimiter.TimeLimiterRegistry;
import org.springframework.cloud.context.config.annotation.RefreshScope;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.env.ConfigurableEnvironment;
import org.springframework.core.env.EnumerablePropertySource;
import org.springframework.core.env.PropertySource;

import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;

@Configuration
public class RoutingConfig {

    // 라우팅 정보 캐싱
    @RefreshScope
    @Bean(name = "routingMap")
    public Map<String, ServiceRouteConfig> routingMap(
            ConfigurableEnvironment env
            , BulkheadRegistry bulkheadRegistry
            , CircuitBreakerRegistry circuitBreakerRegistry
            , TimeLimiterRegistry timeLimiterRegistry
    ) {
        Map<String, ServiceRouteConfig> map = new LinkedHashMap<>();
        // 모든 라우팅 정보를 프로퍼티에서 Map으로 캐싱하는 loop
        for (PropertySource<?> ps : env.getPropertySources()) {
            if (ps.getName().contains("routeInfo") && ps instanceof EnumerablePropertySource<?> eps) {      // -routeInfo 컨피그만 읽도록
                for (String name : eps.getPropertyNames()) {
                    Object rawValue = eps.getProperty(name);
                    String[] parts = rawValue.toString().split(",");

                    String uri = parts[0].trim();
                    long timeout = parts.length > 1 && !parts[1].trim().isBlank() ? Long.parseLong(parts[1].trim()) : 70000L;
                    float failureRate = parts.length > 2 && !parts[2].trim().isBlank() ? Float.parseFloat(parts[2].trim()) : 50.0f;
                    int concurrentSessions = parts.length > 3 && !parts[3].trim().isBlank() ? Integer.parseInt(parts[3].trim()) : 50;

                    ServiceRouteConfig config = new ServiceRouteConfig(uri, timeout, failureRate, concurrentSessions);

                    map.put(name, config);
                }
            }
        }

        // 인스턴스 기동 또는 actuator refresh 될때 모든 라우팅 옵션 초기화(런타임 배포)
        map.forEach((key, config) -> {
            CircuitBreakerConfig cbConfig = CircuitBreakerConfig.custom()       // 서킷브레이커 옵션 (재)조정
                    .failureRateThreshold(config.failureRateThreshold())
                    .slidingWindowType(CircuitBreakerConfig.SlidingWindowType.TIME_BASED)
                    .slidingWindowSize(60)
                    .minimumNumberOfCalls(10)
                    .waitDurationInOpenState(Duration.ofSeconds(10))
                    .build();
            TimeLimiterConfig tlConfig = TimeLimiterConfig.custom()             // 타임아웃 옵션 (재)조정
                    .timeoutDuration(Duration.ofMillis(config.timeoutMs()))
                    .build();

            // 기존 서킷브레이커 + 타임아웃 옵션 제거
            circuitBreakerRegistry.remove(key);
            timeLimiterRegistry.remove(key);
            // 신규 서킷브레이커 + 옵션 적용
            circuitBreakerRegistry.circuitBreaker(key, cbConfig);
            timeLimiterRegistry.timeLimiter(key, tlConfig);

            // 세션제어 옵션 조정 및 런타임 배포
            BulkheadConfig newBulkheadConfig = BulkheadConfig.custom()
                    .maxConcurrentCalls(config.concurrentSessions())
                    .maxWaitDuration(Duration.ofMillis(10))
                    .build();
            bulkheadRegistry.find(key).ifPresentOrElse(
                    existingBulkhead -> existingBulkhead.changeConfig(newBulkheadConfig)        // 세션제어 옵션 재조정(변경)
                    , () -> bulkheadRegistry.bulkhead(key, newBulkheadConfig)                   // 기존 옵션 없을시 새로 조정
            );
        });

        return map;
    }
}

File: EaiGateway/src/main/kotlin/com/yung/cho/eaigateway/config/RoutingConfig.kt
package com.yung.cho.eaigateway.config

import io.github.resilience4j.bulkhead.BulkheadConfig
import io.github.resilience4j.bulkhead.BulkheadRegistry
import io.github.resilience4j.circuitbreaker.CircuitBreakerConfig
import io.github.resilience4j.circuitbreaker.CircuitBreakerRegistry
import io.github.resilience4j.timelimiter.TimeLimiterConfig
import io.github.resilience4j.timelimiter.TimeLimiterRegistry
import org.springframework.cloud.context.config.annotation.RefreshScope
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.env.ConfigurableEnvironment
import org.springframework.core.env.EnumerablePropertySource
import java.time.Duration
import java.util.concurrent.ConcurrentHashMap

@Configuration
class RoutingConfig {

    // 라우팅 정보 캐싱
    @RefreshScope
    @Bean(name = ["routingMap"])
    fun routingMap(
        env: ConfigurableEnvironment,
        bulkheadRegistry: BulkheadRegistry,
        circuitBreakerRegistry: CircuitBreakerRegistry,
        timeLimiterRegistry: TimeLimiterRegistry
    ): Map<String, ServiceRouteConfig> {
        val map = ConcurrentHashMap<String, ServiceRouteConfig>()
        
        // 모든 라우팅 정보를 프로퍼티에서 Map으로 캐싱하는 loop
        for (ps in env.propertySources) {
            if (ps.name.contains("routeInfo") && ps is EnumerablePropertySource<*>) {      // -routeInfo 컨피그만 읽도록
                for (name in ps.propertyNames) {
                    val rawValue = ps.getProperty(name) ?: continue
                    val parts = rawValue.toString().split(",")

                    val uri = parts.getOrNull(0)?.trim().toString()
                    val timeout = parts.getOrNull(1)?.trim()?.takeIf { it.isNotBlank() }?.toLong() ?: 70000L
                    val failureRate = parts.getOrNull(2)?.trim()?.takeIf { it.isNotBlank() }?.toFloat() ?: 50.0f
                    val concurrentSessions = parts.getOrNull(3)?.trim()?.takeIf { it.isNotBlank() }?.toInt() ?: 50

                    val config = ServiceRouteConfig(uri, timeout, failureRate, concurrentSessions)
                    map[name] = config
                }
            }
        }

        // 인스턴스 기동 또는 actuator refresh 될때 모든 라우팅 옵션 초기화(런타임 배포)
        map.forEach { (key, config) ->
            val cbConfig = CircuitBreakerConfig.custom()       // 서킷브레이커 옵션 (재)조정
                .failureRateThreshold(config.failureRateThreshold)
                .slidingWindowType(CircuitBreakerConfig.SlidingWindowType.TIME_BASED)
                .slidingWindowSize(60)
                .minimumNumberOfCalls(10)
                .waitDurationInOpenState(Duration.ofSeconds(10))
                .build()
                
            val tlConfig = TimeLimiterConfig.custom()             // 타임아웃 옵션 (재)조정
                .timeoutDuration(Duration.ofMillis(config.timeoutMs))
                .build()

            val bulkheadConfig = BulkheadConfig.custom()            // 세션제어 옵션 조정
                .maxConcurrentCalls(config.concurrentSessions)
                .maxWaitDuration(Duration.ofMillis(10))
                .build()

            // 기존 서킷브레이커 + 타임아웃 옵션 제거
            circuitBreakerRegistry.remove(key)
            timeLimiterRegistry.remove(key)
            bulkheadRegistry.remove(key)
            
            // 신규 서킷브레이커 + 옵션 적용
            circuitBreakerRegistry.circuitBreaker(key, cbConfig)
            timeLimiterRegistry.timeLimiter(key, tlConfig)
            bulkheadRegistry.bulkhead(key, bulkheadConfig)
        }
        return map
    }
}